File: 0134-gas-station/0134-gas-station.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int currSum = 0;
        int sum = 0;
        int start = 0;
        for (int i = 0; i < gas.size(); i++) {
            currSum += gas[i] - cost[i];
            sum += gas[i] - cost[i];

            if (currSum < 0) {
                start = (i + 1) % gas.size();
                currSum = 0;
            }
        }

        return sum < 0 ? -1 : start;
    }
};

File: 0134-gas-station/0134-gas-station.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        currCost = 0
        currGas = 0
        minDiff = 0

        for i in range(len(gas)):
            currCost += cost[i]
            currGas += gas[i]
            minDiff = min(minDiff, currGas - currCost)
        
        if currCost > currGas:
            return -1
        
        if minDiff >= 0:
            return 0

        currCost = 0
        currGas = 0

        for i in range(len(gas) - 1, -1, -1):
            currCost += cost[i]
            currGas += gas[i]

            if currGas - currCost + minDiff >= 0:
                return i

        return 0